crates/ironrdp-rdpsnd/src/server.rs

```diff
@@ -0,0 +1,209 @@
+    impl_as_any, CompressionCondition, SvcMessage, SvcProcessor, SvcProcessorMessages, SvcServerProcessor,
+};
+use tracing::{debug, error};
+
+use crate::pdu::{self, ClientAudioFormatPdu, QualityMode};
+
+pub type RdpsndSvcMessages = SvcProcessorMessages<RdpsndServer>;
+
+pub trait RdpsndError: std::error::Error + Send + Sync + 'static {}
+
+impl<T> RdpsndError for T where T: std::error::Error + Send + Sync + 'static {}
+
+/// Message sent by the event loop.
+#[derive(Debug)]
+pub enum RdpsndServerMessage {
+    /// Wave data, with timestamp
+    Wave(Vec<u8>, u32),
+    Close,
+    /// Failure received from the OS event loop.
+    ///
+    /// Implementation should log/display this error.
+    Error(Box<dyn RdpsndError>),
+}
+
+pub trait RdpsndServerHandler: Send + std::fmt::Debug {
+    fn get_formats(&self) -> &[pdu::AudioFormat];
+
+    fn start(&mut self, client_format: &ClientAudioFormatPdu) -> Option<u16>;
+
+    fn stop(&mut self);
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum RdpsndState {
+    Start,
+    WaitingForClientFormats,
+    WaitingForQualityMode,
+    WaitingForTrainingConfirm,
+    Ready,
+    Stop,
+}
+
+#[derive(Debug)]
+pub struct RdpsndServer {
+    handler: Box<dyn RdpsndServerHandler>,
+    state: RdpsndState,
+    client_format: Option<ClientAudioFormatPdu>,
+    quality_mode: Option<QualityMode>,
+    block_no: u8,
+    format_no: Option<u16>,
+}
+
+impl RdpsndServer {
+    pub const NAME: ChannelName = ChannelName::from_static(b"rdpsnd\0\0");
+
+    pub fn new(handler: Box<dyn RdpsndServerHandler>) -> Self {
+        Self {
+            handler,
+            state: RdpsndState::Start,
+            client_format: None,
+            quality_mode: None,
+            format_no: None,
+            block_no: 0,
+        }
+    }
+
+    pub fn version(&self) -> PduResult<pdu::Version> {
+        let client_format = self
+            .client_format
+            .as_ref()
+            .ok_or(other_err!("Invalid state - no version"))?;
+
+        Ok(client_format.version)
+    }
+
+    pub fn training_pdu(&mut self) -> PduResult<RdpsndSvcMessages> {
+        let pdu = pdu::TrainingPdu {
+            timestamp: 4231, // a random number
+            data: vec![],
+        };
+        Ok(RdpsndSvcMessages::new(vec![
+            pdu::ServerAudioOutputPdu::Training(pdu).into()
+        ]))
+    }
+
+    pub fn wave(&mut self, data: Vec<u8>, ts: u32) -> PduResult<RdpsndSvcMessages> {
+        let version = self.version()?;
+        let format_no = self.format_no.ok_or(other_err!("Invalid state - no format"))?;
+
+        // the server doesn't wait for wave confirm, apparenyly FreeRDP neither
+        let pdu = if version >= pdu::Version::V8 {
+            pdu::Wave2Pdu {
+                block_no: self.block_no,
+                timestamp: 0,
+                audio_timestamp: ts,
+                format_no,
+                data: data.into(),
+            }
+        } else {
+            unimplemented!()
```

Comment:
issue: Instead of panicking, we should return an error which can be handled gracefully.

Reply:
fixed